src/common/mod.rs

```diff
@@ -44,3 +49,34 @@ pub fn hardware_threads() -> Result<u64, ()> {
         .map_err(|e| debug!("could not parse num cpus from file ({:?}): {}", path, e))
         .map(|i| i + 1)
 }
+
+/// helper function to create a nested map from files with the form of
+/// pkey1 lkey1 lkey2 ... lkeyN
+/// pkey1 value1 value2 ... valueN
+/// pkey2 ...
+pub async fn nested_map_from_file<T: AsRef<Path>>(
+    path: T,
+) -> Result<HashMap<String, HashMap<String, u64>>, std::io::Error> {
+    let mut ret = HashMap::<String, HashMap<String, u64>>::new();
+    let file = File::open(path).await?;
+    let reader = BufReader::new(file);
+    let mut lines = reader.lines();
+    while let Some(keys) = lines.next_line().await? {
+        if let Some(values) = lines.next_line().await? {
+            let keys: Vec<&str> = keys.trim().split_whitespace().collect();
+            let values: Vec<&str> = values.trim().split_whitespace().collect();
+            if keys.len() > 2 {
```

Comment:
why two? shouldn't it be fine if there are only two keys (one pkey and one lkey)?

Reply:
Fixing this. I think I intended to use `>=` here, but this can be rewritten to `if let ...`